src/object_store/src/object/mod.rs

```diff
@@ -105,6 +106,13 @@ pub trait ObjectStore: Send + Sync {
 
     async fn readv(&self, path: &str, block_locs: &[BlockLocation]) -> ObjectResult<Vec<Bytes>>;
 
+    /// Returns a stream that implements `AsyncStream`
+    async fn streaming_read(
+        &self,
+        path: &str,
+        block_loc: Option<BlockLocation>,
+    ) -> ObjectResult<Box<dyn AsyncRead + Unpin>>;
```

Comment:
I would suggest adding a new trait 
```
#[async_trait::async_trait]
trait ObjectStoreOutputStream {
    fn next_bytes(n: usize) -> ObjectResult<Bytes>
}
```
to provide a cleaner abstraction. And the return type of this method will be `ObjectResult<Box<dyn ObjectStoreOutputStream>>`

Reply:
But `AsyncRead` already provides `read_exact(..)` interface, which is quite generic and commonly used in tokio ecosystem.

Reply:
Why not use Stream trait

Reply:
Yes, we could do that. Not sure which one is preferable.

Reply:
> But AsyncRead already provides read_exact(..) interface to read bytes out into a buffer, which is quite generic and commonly used in tokio ecosystem.

Make sense. 

Moreover I think the return type should be `Pin<Box<dyn AsyncRead>` since  we shouldn't force the implementation of the return type to be `Unpin` and also `Pin<Box<...>>` is mostly used in rust async programming.

Reply:
> Why not use Stream trait

Seems that the raw `Stream` trait cannot specified the number of bytes to poll, so `AsyncRead` is more useful than the raw `Stream` trait.